### PR TITLE
Add more `weird.log` entries for malformed TCP options

### PR DESCRIPTION
**Pull request by fox-ds**
When looking at some anomalous traffic and diving into the `TCPSessionAdapter.cc` file, it appears that Zeek just assumes that the TCP options are always according to the RFC. However, in some cases, TCP options can even be misused for covert C2 channels. In this example, we were looking at a `ls -la` command that was embedded into a TCP option.

After replaying the pcap in Zeek and looking at the `tcp_option` and `tcp_options` events, there wasn't any sign of something extraordinary about the TCP options. The comments in the `TCPSessionAdapter.cc` file hint at possible future logging entries for these anomalous options, which are implemented in this commit.

In summary, this commit adds a few new entries in the `weird.log`:

- tcp_nonstandard_option
- tcp_option_no_room_for_length
- tcp_option_trashed_length
- tcp_option_length_exceeds_options_space
- tcp_remaining_options_after_TCPOPT_EOL

Note that this change will only add these new entries in the `weird.log` when either the `tcp_option` or `tcp_options` event is overridden.

### CODE REVIEW

**Review by 0xxon — COMMENTED** [review]
This generally looks good to me...

However, this also changes some low-level behavior in our TCP stack. As such, I think this is something that optimally should see quite a bit of testing. I know that in principle this should just work - but I am always worried that there is something that will break because of this.

One slightly concerning change to me in this is the fact that we continue parsing after `TCPOPT_EOL`. I think it is a good idea to do that - but I am not sure what that might cause in practice (even if it is just excessive warnings because something adds random padding afterwards).

The second thing that I don't quite get is the option length patching. Why is this desirable, and are we sure that it is ok to cast away the const-ness of the tcphdr (I only tracked it up one function call where it also seems to be const).

File: src/packet_analysis/protocol/tcp/TCPSessionAdapter.cc [hunk 1]
```diff
@@ -1670,29 +1670,66 @@ int TCPSessionAdapter::ParseTCPOptions(const struct tcphdr* tcp, bool is_orig)
 
 		unsigned int opt_len;
 
+		// Non-standard TCP option kinds should not be common
+		// https://www.iana[.]org/assignments/tcp-parameters/tcp-parameters.xhtml
```
Comment (grigorescu) [inline src/packet_analysis/protocol/tcp/TCPSessionAdapter.cc:1674]: Minor nit, but I don't understand why this URL is sanitized. Just makes it harder to actually reference.

File: src/packet_analysis/protocol/tcp/TCPSessionAdapter.cc [hunk 2]
```diff
@@ -1670,29 +1670,66 @@ int TCPSessionAdapter::ParseTCPOptions(const struct tcphdr* tcp, bool is_orig)
 
 		unsigned int opt_len;
 
+		// Non-standard TCP option kinds should not be common
```
[the remark this answers is not in the document]
Comment (fox-ds) [reply src/packet_analysis/protocol/tcp/TCPSessionAdapter.cc:1673]: Good suggestion, I'll change this!

File: src/packet_analysis/protocol/tcp/TCPSessionAdapter.cc (hunk 1 above)
Comment (fox-ds) [reply src/packet_analysis/protocol/tcp/TCPSessionAdapter.cc:1674]: It's a force of habit, I guess. I'll remove it here.

**Comment by fox-ds** [comment]
@0xxon Thanks for your feedback! I'll have a look at the option length patching again, and check if this is really needed or can be left out of this PR. I primarily did this to get the command data in the `tcp_option` event, but this solution may not be desirable. 

Regarding the additional data after the `TCPOPT_EOL`; maybe we could only send out a Weird log if the next byte is a valid TCP option kind. I'll have a look at the feedback in the coming weeks, thanks again!

**Comment by 0xxon** [comment]
That sounds great. Let us know if you need anything in the meantime :)

**Comment by rsmmr** [comment]
@fox-ds fyi, we'll start wrapping things up for 5.0 soon if you'd like this to still go in.

**Comment by rsmmr** [comment]
I'm hearing that more changes are coming here but won't make it for 5.0. I'll close this for now, please reopen when updated.
